File: go/tests/other/mpr121_test/main.go
//go:build linux && !tinygo

// MPR121 hardware test — Linux host.
package main

import (
	"fmt"
	"os"
	"strconv"

	"github.com/tuhde/Periph/go/periph/chips/other"
	"github.com/tuhde/Periph/go/periph/connection"
)

func main() {
	bus, err := strconv.Atoi(envOr("I2C_BUS", "1"))
	if err != nil {
		fmt.Fprintln(os.Stderr, "I2C_BUS:", err)
		os.Exit(2)
	}
	addr, err := strconv.ParseUint(envOr("I2C_ADDR", "0x5A"), 0, 8)
	if err != nil {
		fmt.Fprintln(os.Stderr, "I2C_ADDR:", err)
		os.Exit(2)
	}

	conn, err := connection.NewI2CConnection(bus, uint8(addr), nil, nil)
	if err != nil {
		fmt.Fprintln(os.Stderr, "connection:", err)
		os.Exit(2)
	}
	defer conn.Close()

	chip, err := other.NewMPR121Full(conn)
	if err != nil {
		fmt.Fprintln(os.Stderr, "new:", err)
		os.Exit(2)
	}

	passed := 0
	failed := 0
	check := func(label string, ok bool) {
		if ok {
			fmt.Printf("PASS %s\n", label)
			passed++
		} else {
			fmt.Printf("FAIL %s\n", label)
			failed++
		}
	}

	t, err := chip.Touched()
	check("touched in 0..4095", err == nil && t <= 0xFFF)

	if _, err := chip.IsTouched(0); err == nil {
		check("is_touched returns bool", true)
	}

	f0, err := chip.Filtered(0)
	check("filtered(0) in 0..1023", err == nil && f0 <= 1023)

	b0, err := chip.Baseline(0)
	check("baseline(0) in 0..1023", err == nil && b0 <= 1023)

	oor, err := chip.OORStatus()
	check("oor_status in 0..8191", err == nil && oor <= 0x1FFF)

	if err := chip.Stop(); err == nil {
		check("stop accepted", true)
	}
	if err := chip.ConfigureThresholds(0, 15, 8); err == nil {
		check("configure_thresholds accepted", true)
	}
	if err := chip.ConfigureAllThresholds(12, 6); err == nil {
		check("configure_all_thresholds accepted", true)
	}
	if err := chip.ConfigureProximityThresholds(8, 4); err == nil {
		check("configure_proximity_thresholds accepted", true)
	}
	if err := chip.ConfigureBaselineFilter(1, 1, 0, 0, 1, 1, 0, 0, 1, 0, 0); err == nil {
		check("configure_baseline_filter accepted", true)
	}
	if err := chip.ConfigureSampling(16, 1, 0, 0, 4); err == nil {
		check("configure_sampling accepted", true)
	}
	if err := chip.ConfigureDebounce(1, 1); err == nil {
		check("configure_debounce accepted", true)
	}
	if err := chip.ConfigureAutoconfig(3300, 0, false, true, true); err == nil {
		check("configure_autoconfig accepted", true)
	}
	if err := chip.Start(12, 2, 0); err == nil {
		check("start accepted", true)
	}
	if err := chip.EnableInterrupt(other.SOURCE_OOR); err == nil {
		check("enable_interrupt accepted", true)
	}
	if err := chip.DisableInterrupt(other.SOURCE_OOR); err == nil {
		check("disable_interrupt accepted", true)
	}
	if err := chip.ClearOvercurrent(); err == nil {
		check("clear_overcurrent accepted", true)
	}
	if err := chip.Reset(); err == nil {
		check("reset accepted", true)
	}

	fmt.Printf("===DONE: %d passed, %d failed===\n", passed, failed)
	if failed > 0 {
		os.Exit(1)
	}
}

func envOr(k, def string) string {
	if v, ok := os.LookupEnv(k); ok {
		return v
	}
	return def
}

File: rust/tests/other/mpr121_test/src/main.rs
use linux_embedded_hal::{Delay, I2cdev};
use periph::chips::other::{Mpr121Full, SOURCE_OOR};
use std::thread::sleep;
use std::time::Duration;

fn check(label: &str, cond: bool) {
    if cond { println!("PASS {}", label); }
    else    { println!("FAIL {}", label); }
}

fn main() {
    let i2c_bus: u8 = std::env::var("I2C_BUS").ok().and_then(|v| v.parse().ok()).unwrap_or(1);
    let addr: u8 = std::env::var("I2C_ADDR")
        .ok()
        .and_then(|v| u8::from_str_radix(v.trim_start_matches("0x"), 16).ok())
        .unwrap_or(0x5A);

    let dev = I2cdev::new(format!("/dev/i2c-{}", i2c_bus)).expect("open i2c bus"); // Open I²C bus, (/dev/i2c-N) → I2cdev
    let mut delay = Delay;
    let mut chip = Mpr121Full::new(dev, addr, &mut delay).expect("init");         // Construct MPR121 Full, (i2c, addr=0x5A, delay) → Result<Mpr121Full, _>

    let t = chip.touched().expect("touched");
    check("touched in 0..4095", t <= 0xFFF);

    let f0 = chip.filtered(0).expect("filtered");
    check("filtered(0) in 0..1023", f0 <= 1023);

    let b0 = chip.baseline(0).expect("baseline");
    check("baseline(0) in 0..1023", b0 <= 1023);

    let oor = chip.oor_status().expect("oor");
    check("oor_status in 0..8191", oor <= 0x1FFF);

    chip.stop().expect("stop");
    chip.configure_thresholds(0, 15, 8).expect("thresholds");
    chip.configure_all_thresholds(12, 6).expect("all thresholds");
    chip.configure_proximity_thresholds(8, 4).expect("prox thresholds");
    chip.configure_baseline_filter(1, 1, 0, 0, 1, 1, 0, 0, 1, 0, 0).expect("baseline filter");
    chip.configure_sampling(16, 1, 0, 0, 4).expect("sampling");
    chip.configure_debounce(1, 1).expect("debounce");
    chip.configure_autoconfig(3300, 0, false, true, true).expect("autoconfig");
    check("configuration methods accepted", true);

    chip.enable_interrupt(SOURCE_OOR).expect("enable int");
    chip.disable_interrupt(SOURCE_OOR).expect("disable int");
    chip.clear_overcurrent().expect("clear ovrc");
    check("interrupt API accepted", true);

    chip.reset(&mut delay).expect("reset");
    check("reset completed", true);

    println!("===DONE===");
    let _ = sleep(Duration::from_millis(10));
}
